### PR TITLE
refactor(web): move timeline-manager ops back into month-group; cleanup API

### PR DESCRIPTION
**Pull request by midzelis**
Consolidates asset operation logic within TimelineManager class and removes the now redundant  operations-support.svelte.ts file. 

Combines addAsset/updateAsset to be upsertAsset. 

Changes:
- Move `addAssetsToMonthGroups` logic into TimelineManager's `addAssetsToSegments`, `upsertAssetIntoSegment`, `postCreateSegments`, and `postUpsert` methods
- Move `runAssetOperation` from operations-support into TimelineManager's private `#runAssetOperation` method
- Rename public `addAssets`/`updateAssets` methods to unified `upsertAssets` for consistency
- Delete internal/operations-support.svelte.ts
- Update WebsocketSupport to use `upsertAssets` for both add and update operations
- Fix AssetOperation return type to allow undefined/void operations (not just `{ remove: boolean }`)
- Update MonthGroup constructor to accept `loaded` parameter for better initialization control
- Update all test references from `addAssets`/`updateAssets` to `upsertAssets`

This refactoring improves code maintainability by eliminating duplicate logic and consolidating all asset operations within the TimelineManager class where they belong.

Next PR: #23577

### CODE REVIEW

File: web/src/lib/managers/timeline-manager/day-group.svelte.ts [hunk 1]
```diff
@@ -122,7 +122,11 @@ export class DayGroup {
 
       const asset = this.viewerAssets[index].asset!;
       const oldTime = { ...asset.localDateTime };
-      let { remove } = operation(asset);
+      const opResult = operation(asset);
+      let remove = false;
+      if (opResult) {
+        remove = (opResult as { remove: boolean }).remove ?? false;
+      }
```
Comment (mertalev) [inline web/src/lib/managers/timeline-manager/day-group.svelte.ts:136]: This seems like an odd API. Does the removal decision actually depend on the individual asset in practice, or is it already known when the operation is made whether the asset should be removed?

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts [hunk 2]
```diff
@@ -400,19 +404,16 @@ export class TimelineManager extends VirtualScrollManager {
...
+    return this.#runAssetOperation(new Set(ids), operation);
+  }
+
+  #updateAssets(assets: TimelineAsset[]) {
+    // eslint-disable-next-line svelte/prefer-svelte-reactivity
+    const lookup = new Map<string, TimelineAsset>(assets.map((asset) => [asset.id, asset]));
+    // eslint-disable-next-line svelte/prefer-svelte-reactivity
+    const { unprocessedIds } = this.#runAssetOperation(new Set(lookup.keys()), (asset) =>
+      updateObject(asset, lookup.get(asset.id)),
     );
```
Comment (mertalev) [inline web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts]: This is making at least 3 copies of the data. Is all of this really necessary?

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts [hunk 3]
```diff
@@ -422,17 +423,83 @@ export class TimelineManager extends VirtualScrollManager {
...
-    const { unprocessedIds } = runAssetOperation(
-      this,
-      new SvelteSet(ids),
-      () => {
-        return { remove: true };
-      },
-      { order: this.#options.order ?? AssetOrder.Desc },
-    );
+    // eslint-disable-next-line svelte/prefer-svelte-reactivity
+    const { unprocessedIds } = this.#runAssetOperation(new Set(ids), () => ({ remove: true }));
```
Comment (mertalev) [inline web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts]: Yeah so this empty closure with a return value seems kind of weird. I'm also not seeing why it needs to make a set here given it's already making it into a SvelteSet later.

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts [hunk 4]
```diff
@@ -422,17 +423,83 @@ export class TimelineManager extends VirtualScrollManager {
...
+    if (combinedMoveAssets.length > 0) {
+      this.addAssetsToSegments(combinedMoveAssets.flat().map((a) => a.asset));
+    }
+    const changedGeometry = changedMonthGroups.size > 0;
+    for (const month of changedMonthGroups) {
+      updateGeometry(this, month, { invalidateHeight: true });
+    }
+    if (changedGeometry) {
+      this.updateIntersections();
+    }
```
Comment (mertalev) [inline web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts:510]: addAssetsToSegments sorts and updates geometry as well, so is it duplicating work? Also, I'm not a big fan of how it's sorting even when it's getting sorted assets from the server.

File: web/src/lib/managers/timeline-manager/day-group.svelte.ts (hunk 1 above)
Comment (midzelis) [reply web/src/lib/managers/timeline-manager/day-group.svelte.ts:136]: The idea is that the operation is generic enough to be customized by each subclass of VirtualScrollManager. In the timeline case - no - changing the date of an asset may cause it to be removed from one day, and put into another day-group, or even a different month-group. 

However, I do think the return type of the operation (` (asset: TimelineAsset) => { remove: boolean } | unknown;`) is awkward. For the 'base class' of the operation, I'll just have it return 'unknown' - and use the subclass covariant return types add things like {remove: boolean} 

Perhaps the naming could be better, like - "updatedAssetLeftGroupAndMustBeReadded" - I'll think on it

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts (hunk 2 above)
Comment (midzelis) [reply web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts]: I've adjusted this to minimize the copies - and instead of using maps and arrows, to use for..of. 

The lookup map is still necessary - otherwise, an array.find() would be required inside the operation which would be slower than the map. 

I removed the new Set(lookup.keys()) - which was a defensive copy (so that mutations on the param wouldn't harm the overall algorithm). 

I was able to remove the MoveAsset[][] and the flat/map ops on it (and the MoveAsset type completely) and just use TimelineAsset[] instead.

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts (hunk 3 above)
Comment (midzelis) [reply web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts]: Removed this defensive copy - and the return value was unused, so removed that too

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts (hunk 4 above)
Comment (midzelis) [reply web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts:510]: 1. addAssetsToSegments will sort/updateGeo for the target day-group and months. 
2. changedMonthGroups will update geometry for the months/day-groups that have assets leave. 

That being said, there could be some logic flaws in here - I'll add a few tests to make sure duplicate work won't be done. 

For your last point - yes - the timeline gets pre-sorted items from the server on initial load. However, the websocket listener can also update assets - change their dates, add assets, remove assets. Additionally, menu/keyboard actions can do the same thing. (changing the date in the UI, adding/removing, etc) Which is why the timeline manager also needs to sort. (It tries to only sort when necessary - it changing geometry, or assets moving inbetween day/month groups) This is why this operation is relatively complex - it tries to do as many things in bulk, and only do the minimum amount of things required.

File: web/src/lib/utils/timeline-util.ts [hunk 5]
```diff
@@ -222,12 +221,35 @@ export const plainDateTimeCompare = (ascending: boolean, a: TimelineDateTime, b:
...
+  // Check if native Set.prototype.difference is available (ES2025)
+  const setWithDifference = setA as unknown as Set<T> & { difference?: (other: Set<T>) => Set<T> };
+
+  if (setWithDifference.difference && typeof setWithDifference.difference === 'function') {
+    try {
+      // Use native difference method if available
+      return setWithDifference.difference(setB);
+    } catch {
+      // Fall through to manual implementation if native method fails
+    }
```
Comment (mertalev) [inline web/src/lib/utils/timeline-util.ts]: The try/catch seems unnecessary

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts (hunk 4 above)
Comment (mertalev) [reply web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts:510]: The ideal would be for the initial bucket load to avoid sorting - there were sizable gains to this when I tested it when we changed to columnar bucket responses. Everything after that (like websockets) can sort without issue. Do you think that's feasible to allow with this API?

File: web/src/lib/utils/timeline-util.ts (hunk 5 above)
Comment (midzelis) [reply web/src/lib/utils/timeline-util.ts]: set.difference only works on 86% of browsers: https://caniuse.com/?search=set.difference

Comment (mertalev) [reply web/src/lib/utils/timeline-util.ts]: Sure, but there's already a check if it exists

File: web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts (hunk 4 above)
Comment (midzelis) [reply web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts:510]: Yes, we can make the load from server not sort. However, that is a different code path. This code path is only for mutations (add/update/remove) after the initial load.

Comment (mertalev) [reply web/src/lib/managers/timeline-manager/timeline-manager.svelte.ts:510]: Ah, gotcha. That can be handled separately from this PR in that case

**Review by mertalev — APPROVED** [review]
Code LGTM

File: web/src/lib/managers/timeline-manager/day-group.svelte.ts [hunk 6]
```diff
@@ -31,6 +31,7 @@ export class DayGroup {
     this.monthGroup = monthGroup;
     this.day = day;
     this.groupTitle = groupTitle;
+    onCreateDayGroup(this);
```
Comment (mertalev) [inline web/src/lib/managers/timeline-manager/day-group.svelte.ts]: Something neat I learned: if you wrap this in [$inspect](https://svelte.dev/docs/svelte/$inspect), it gets completely removed by the preprocessor in prod builds.

Comment (midzelis) [reply web/src/lib/managers/timeline-manager/day-group.svelte.ts]: $inspect() is only valid within the lifecycle of a component initialization. This unit test uses TimelineManager.svelte directly, without a component - so it fails. Would have been nice though.

Comment (mertalev) [reply web/src/lib/managers/timeline-manager/day-group.svelte.ts]: The recommendation from the Svelte folks is to use `if (import.meta.env.DEV)` for tree-shaking. It's apparently a Vite thing.

Comment (midzelis) [reply web/src/lib/managers/timeline-manager/day-group.svelte.ts]: very nice - updated.

**Comment by jrasm91** [comment]
When an asset is unarchived from the /archive page it isn't removed from the timeline. Can you please add a test for this?

**Comment by midzelis** [comment]
Going to split this one up too
